src/storage/src/render/upsert.rs

```diff
@@ -298,13 +308,23 @@ where
                 }
             }
 
-            state
+            match state
                 .merge_snapshot_chunk(
                     events.drain(..),
                     PartialOrder::less_equal(&resume_upper, &snapshot_upper),
                 )
                 .await
-                .expect("hashmap impl to not fail");
+            {
+                Ok(_) => {}
+                Err(e) => {
+                    let update = get_health_update(e, "failure while rehydrating state");
+                    health_output
+                        .give(&health_cap, (source_config.worker_id, 0, update))
+                        .await;
+                    std::future::pending::<()>().await;
+                    unreachable!("pending future never returns");
+                }
```

Comment:
can we factor this out into a function like `async fn process_upsert_state_error(e: Error, health_output: ...)`? We do the same thing 3 times, right?

Reply:
Yeah, this is what I was having trouble with. I was trying to do a `map` result rather unsuccessfully. Let me give it a shot again.

Reply:
let me know if you need help!

Reply:
done!